### ISSUE TITLE
1 of 3 nodes failing with "no recovery target specified" 

### ISSUE THREAD

**Issue by ktechkez** [zalando/patroni#736]
Hi, 

I have been struggling for a couple of days now getting a 3 node setup with patroni/postgres running properly. I've been following this guide: https://www.linode.com/docs/databases/postgresql/create-a-highly-available-postgresql-cluster-using-patroni-and-haproxy/. 

The master and a single secondary node start fine every time, but a third node always fails to start postgres with the following: 

`Jul 16 14:30:45 devpostgres03 patroni[10951]: 2018-07-16 14:30:45,213 INFO: Lock owner: devpostgres01; I am devpostgres03
Jul 16 14:30:45 devpostgres03 patroni[10951]: 2018-07-16 14:30:45,228 INFO: Lock owner: devpostgres01; I am devpostgres03
Jul 16 14:30:45 devpostgres03 patroni[10951]: 2018-07-16 14:30:45,255 INFO: Local timeline=14 lsn=0/60005F8
Jul 16 14:30:45 devpostgres03 patroni[10951]: 2018-07-16 14:30:45,269 INFO: master_timeline=22
Jul 16 14:30:45 devpostgres03 patroni[10951]: 2018-07-16 14:30:45,271 INFO: master: history=1        0/19C4218        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 2        0/19C4638        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 3        0/40000D0        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 4        0/4000410        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 5        0/4000830        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 6        0/4000C50        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 7        0/4000F90        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 8        0/40011F0        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 9        0/4001370        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 10        0/40015D0        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 11        0/6000178        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 12        0/60002F8        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 13        0/6000478        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 14        0/60007B8        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 15        0/6000938        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 16        0/80000D0        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 17        0/8000330        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 18        0/80004B0        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 19        0/8000630        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 20        0/8000970        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 21        0/8000AF0        no recovery target specified
Jul 16 14:30:45 devpostgres03 patroni[10951]: 2018-07-16 14:30:45,280 INFO: starting as a secondary
Jul 16 14:30:46 devpostgres03 patroni[10951]: 2018-07-16 14:30:46,004 INFO: postmaster pid=11404
Jul 16 14:30:46 devpostgres03 patroni[10951]: < 2018-07-16 14:30:46.008 UTC >LOG:  redirecting log output to logging collector process
Jul 16 14:30:46 devpostgres03 patroni[10951]: < 2018-07-16 14:30:46.008 UTC >HINT:  Future log output will appear in directory "pg_log".
Jul 16 14:30:46 devpostgres03 patroni[10951]: 10.10.10.213:5432 - no response
Jul 16 14:30:47 devpostgres03 patroni[10951]: 2018-07-16 14:30:47,020 ERROR: postmaster is not running
Jul 16 14:30:47 devpostgres03 patroni[10951]: 2018-07-16 14:30:47,056 INFO: Lock owner: devpostgres01; I am devpostgres03
Jul 16 14:30:47 devpostgres03 patroni[10951]: 2018-07-16 14:30:47,061 INFO: failed to start postgres
`

Here is my configuration: 

`scope: postgres
namespace: /db/
name: devpostgres03

restapi:
    listen: 10.10.10.213:8008
    connect_address: 10.10.10.213:8008
etcd:
    host: 10.10.10.214:2379

bootstrap:
    dcs:
        ttl: 30
        loop_wait: 10
        retry_timeout: 10
        maximum_lag_on_failover: 1048576
        postgresql:
            use_pg_rewind: true
            recovery_conf: 
                recovery_target: immediate
    initdb:
    - encoding: UTF8
    - data-checksums

    pg_hba:
    - host replication replicator 127.0.0.1/32 md5
    - host replication replicator 10.10.10.211/0 md5
    - host replication replicator 10.10.10.213/0 md5
    - host replication replicator 10.10.10.212/0 md5
    - host all all 0.0.0.0/0 md5

    users:
        admin:
            password: admin
            options:
                - createrole
                - createdb

postgresql:
    listen: 10.10.10.213:5432
    connect_address: 10.10.10.213:5432
    data_dir: /data/patroni
    pgpass: /tmp/pgpass
    authentication:
        replication:
            username: replicator
            password: rep-pass
        superuser:
            username: postgres
            password: secretpassword
    parameters:
        unix_socket_directories: '.'

tags:
    nofailover: false
    noloadbalance: false
    clonefrom: false
    nosync: false`

And finally, I am starting via systemd with the following service file: 

`[Unit]
Description=Patroni runners to orchestrate a high-availability PostgreSQL
After=syslog.target network.target

[Service]
Type=simple
User=postgres
Group=postgres
ExecStart=/usr/bin/patroni /etc/patroni/patroni.yml
KillMode=process
TimeoutSec=30
Restart=no

[Install]
WantedBy=multi-user.target`

I tried specifing the recovery_target key in my patroni config but I still receive this issue constantly on one of the slaves. 
Any help is hugely appreciated as I've been struggling for a while on this. 

Thanks, 

Carey

**Comment by alexeyklyukin** [comment]
```
Jul 16 14:30:46 devpostgres03 patroni[10951]: 10.10.10.213:5432 - no response
Jul 16 14:30:47 devpostgres03 patroni[10951]: 2018-07-16 14:30:47,020 ERROR: postmaster is not running
Jul 16 14:30:47 devpostgres03 patroni[10951]: 2018-07-16 14:30:47,056 INFO: Lock owner: devpostgres01; I am devpostgres03
Jul 16 14:30:47 devpostgres03 patroni[10951]: 2018-07-16 14:30:47,061 INFO: failed to start postgres
```

According to the lines above Patroni failed to start Postgres. Can you look at the Postgres logs for the devpostgres03 (you may need to enable them by setting log_destionation='csvlog' and logging_collector='on` at postgresql/parameters (right below the "unix_socket_directory") in Patroni configuration and then restarting Patroni on the node.

**Comment by ktechkez** [comment]
Hi Alex thanks for quick response. Here is the contents of the csv output on the node that isn't working: 

`2018-07-17 10:17:34.759 UTC,,,10558,,5b4dc23e.293e,1,,2018-07-17 10:17:34 UTC,,0,LOG,00000,"ending log output to stderr",,"Future log output will go to log destination ""csvlog"".",,,,,,,""
2018-07-17 10:17:34.762 UTC,,,10561,,5b4dc23e.2941,1,,2018-07-17 10:17:34 UTC,,0,LOG,00000,"database system was shut down in recovery at 2018-07-16 13:43:34 UTC",,,,,,,,,""
2018-07-17 10:17:34.762 UTC,,,10561,,5b4dc23e.2941,2,,2018-07-17 10:17:34 UTC,,0,LOG,00000,"entering standby mode",,,,,,,,,""
2018-07-17 10:17:34.762 UTC,,,10561,,5b4dc23e.2941,3,,2018-07-17 10:17:34 UTC,,0,LOG,00000,"invalid resource manager ID in primary checkpoint record",,,,,,,,,""
2018-07-17 10:17:34.762 UTC,,,10561,,5b4dc23e.2941,4,,2018-07-17 10:17:34 UTC,,0,PANIC,XX000,"could not locate a valid checkpoint record",,,,,,,,,""
2018-07-17 10:17:34.763 UTC,,,10558,,5b4dc23e.293e,2,,2018-07-17 10:17:34 UTC,,0,LOG,00000,"startup process (PID 10561) was terminated by signal 6: Aborted",,,,,,,,,""
2018-07-17 10:17:34.763 UTC,,,10558,,5b4dc23e.293e,3,,2018-07-17 10:17:34 UTC,,0,LOG,00000,"aborting startup due to startup process failure",,,,,,,,,""
2018-07-17 10:17:37.909 UTC,,,10617,,5b4dc241.2979,1,,2018-07-17 10:17:37 UTC,,0,LOG,00000,"ending log output to stderr",,"Future log output will go to log destination ""csvlog"".",,,,,,,""
2018-07-17 10:17:37.911 UTC,,,10620,,5b4dc241.297c,1,,2018-07-17 10:17:37 UTC,,0,LOG,00000,"database system was shut down in recovery at 2018-07-16 13:43:34 UTC",,,,,,,,,""
2018-07-17 10:17:37.912 UTC,,,10620,,5b4dc241.297c,2,,2018-07-17 10:17:37 UTC,,0,LOG,00000,"entering standby mode",,,,,,,,,""
2018-07-17 10:17:37.912 UTC,,,10620,,5b4dc241.297c,3,,2018-07-17 10:17:37 UTC,,0,LOG,00000,"invalid resource manager ID in primary checkpoint record",,,,,,,,,""
2018-07-17 10:17:37.912 UTC,,,10620,,5b4dc241.297c,4,,2018-07-17 10:17:37 UTC,,0,PANIC,XX000,"could not locate a valid checkpoint record",,,,,,,,,""
2018-07-17 10:17:37.912 UTC,,,10617,,5b4dc241.2979,2,,2018-07-17 10:17:37 UTC,,0,LOG,00000,"startup process (PID 10620) was terminated by signal 6: Aborted",,,,,,,,,""
2018-07-17 10:17:37.912 UTC,,,10617,,5b4dc241.2979,3,,2018-07-17 10:17:37 UTC,,0,LOG,00000,"aborting startup due to startup process failure",,,,,,,,,""
2018-07-17 10:17:47.371 UTC,,,10641,,5b4dc24b.2991,1,,2018-07-17 10:17:47 UTC,,0,LOG,00000,"ending log output to stderr",,"Future log output will go to log destination ""csvlog"".",,,,,,,""
2018-07-17 10:17:47.374 UTC,,,10644,,5b4dc24b.2994,1,,2018-07-17 10:17:47 UTC,,0,LOG,00000,"database system was shut down in recovery at 2018-07-16 13:43:34 UTC",,,,,,,,,""
2018-07-17 10:17:47.374 UTC,,,10644,,5b4dc24b.2994,2,,2018-07-17 10:17:47 UTC,,0,LOG,00000,"entering standby mode",,,,,,,,,""
2018-07-17 10:17:47.374 UTC,,,10644,,5b4dc24b.2994,3,,2018-07-17 10:17:47 UTC,,0,LOG,00000,"invalid resource manager ID in primary checkpoint record",,,,,,,,,""
2018-07-17 10:17:47.374 UTC,,,10644,,5b4dc24b.2994,4,,2018-07-17 10:17:47 UTC,,0,PANIC,XX000,"could not locate a valid checkpoint record",,,,,,,,,""
2018-07-17 10:17:47.375 UTC,,,10641,,5b4dc24b.2991,2,,2018-07-17 10:17:47 UTC,,0,LOG,00000,"startup process (PID 10644) was terminated by signal 6: Aborted",,,,,,,,,""
2018-07-17 10:17:47.375 UTC,,,10641,,5b4dc24b.2991,3,,2018-07-17 10:17:47 UTC,,0,LOG,00000,"aborting startup due to startup process failure",,,,,,,,,""
2018-07-17 10:17:57.370 UTC,,,10674,,5b4dc255.29b2,1,,2018-07-17 10:17:57 UTC,,0,LOG,00000,"ending log output to stderr",,"Future log output will go to log destination ""csvlog"".",,,,,,,""
2018-07-17 10:17:57.372 UTC,,,10677,,5b4dc255.29b5,1,,2018-07-17 10:17:57 UTC,,0,LOG,00000,"database system was shut down in recovery at 2018-07-16 13:43:34 UTC",,,,,,,,,""
2018-07-17 10:17:57.372 UTC,,,10677,,5b4dc255.29b5,2,,2018-07-17 10:17:57 UTC,,0,LOG,00000,"entering standby mode",,,,,,,,,""
2018-07-17 10:17:57.372 UTC,,,10677,,5b4dc255.29b5,3,,2018-07-17 10:17:57 UTC,,0,LOG,00000,"invalid resource manager ID in primary checkpoint record",,,,,,,,,""
2018-07-17 10:17:57.372 UTC,,,10677,,5b4dc255.29b5,4,,2018-07-17 10:17:57 UTC,,0,PANIC,XX000,"could not locate a valid checkpoint record",,,,,,,,,""
2018-07-17 10:17:57.373 UTC,,,10674,,5b4dc255.29b2,2,,2018-07-17 10:17:57 UTC,,0,LOG,00000,"startup process (PID 10677) was terminated by signal 6: Aborted",,,,,,,,,""
2018-07-17 10:17:57.373 UTC,,,10674,,5b4dc255.29b2,3,,2018-07-17 10:17:57 UTC,,0,LOG,00000,"aborting startup due to startup process failure",,,,,,,,,""
`

**Comment by alexeyklyukin** [comment]
```
2018-07-17 10:17:34 UTC,,0,LOG,00000,"invalid resource manager ID in primary checkpoint record

 2018-07-17 10:17:34.762 UTC,,,10561,,5b4dc23e.2941,4,,2018-07-17 10:17:34 UTC,,0,PANIC,XX000,"could not locate a valid checkpoint record
```

Looks like your replica has not being initialized properly (perhaps something happened during the basebackup that made it abort prematurely). You can run  
`patronictl -c patroni.yaml reinit postgres devpostgres03` to reinitialize this node and see whether it would fix the issue.

I also noticed there is no archive_command and restore_command in your configuration files. This means that if the replica falls out-of-date while streaming your changes (because, for instance, it cannot keep up with the ratio of changes or there is a network issue and the master has already recycled the WAL segments still required for the replica) it won't be able to continue without reinitializing. However, it looks like your issue is not caused by this, as in the case of the missing WAL segments the error message will be different.

**Comment by ktechkez** [comment]
Interesting, I just tried this and I received the following errorL

patronictl -c /etc/patroni/patroni.yaml reinit postgres devpostgres03
Error: 'Can not find suitable configuration of distributed configuration store\nAvailable implementations: kubernetes, consul, zookeeper, exhibitor, etcd'

I received this error previously with patroni when I was missing the dependencies but I'm installing them all via pip as part of the deployment script. Any reason why patronictl command would still be causing this?

**Comment by alexeyklyukin** [comment]
Did you supply the correct filename? Your init script mentions /etc/patroni/patroni.yml (not yaml) (patronictll could also be more verbose if you give it a non-existing filename)

**Comment by ktechkez** [comment]
D'oh! Silly me, copied your command, added the path but missed the wrong file extension. Looks like that did the trick! Current status: 

patronictl -c /etc/patroni/patroni.yml list postgres
+----------+---------------+--------------+--------+---------+-----------+
| Cluster  |     Member    |     Host     |  Role  |  State  | Lag in MB |
+----------+---------------+--------------+--------+---------+-----------+
| postgres | devpostgres01 | 10.10.10.211 | Leader | running |       0.0 |
| postgres | devpostgres02 | 10.10.10.212 |        | running |       0.0 |
| postgres | devpostgres03 | 10.10.10.213 |        | running |       0.0 |

Thanks ever so much!

**Comment by ktechkez** [comment]
While I'm here, would you be able to give me an example of the archive and recovery command settings I should use? Thankyou!

**Comment by sysadmin-james** [comment]
hi there,
i was getting following error on one of my replicas  "master: history=1	0/678BA08	no recovery target specified"

so i have tried the suggested action on this issue with the following command
docker exec -it  UAT_postgres2.1.z7i0uuwifizmap7rhgq900fvx   patronictl -c patroni.yml reinit postgres postgres2

this gave following output/error
+---------+--------+------+------+-------+----+-----------+
| Cluster | Member | Host | Role | State | TL | Lag in MB |
+---------+--------+------+------+-------+----+-----------+
+---------+--------+------+------+-------+----+-----------+

**_Error: postgres2 is not a member of cluster_**

it says postgres2 is not a member of cluster even though it is part of the cluster. 
Any reason why patroni is suggesting postgres2 is not part of the cluster.

Here is the list of nodes/dbs part of of cluster

docker exec -it UAT_postgres2.1.z7i0uuwifizmap7rhgq900fvx  patronictl -c patroni.yml list
+---------------------+--------------+--------------+--------+---------+----+-----------+-----------------+------------------|
|       Cluster       |    Member    |     Host                |  Role   |  State   | TL | Lag in MB | Pending restart |
+---------------------+--------------+--------------+--------+---------+----+----------- +----------------+------------------|
| postgres-cluster | postgres1 | postgres1 |                         | running | 10 |         0         |        *             |
| postgres-cluster | postgres2 | postgres2 |                         | running | 10 |       650       |        *             |
| postgres-cluster | postgres3 | postgres3 |             Leader | running | 10 |         0         |        *             |
+---------------------+--------------+--------------+--------+---------+----+-----------+-----------------+------------------|

also postgres2  node/service lags with rest of the cluster by big 650 MB

A quick responce will be appreciated.
Thanks & regards
sysadmin-james

**Comment by CyberDem0n** [comment]
> i was getting following error on one of my replicas "master: history=1 0/678BA08 no recovery target specified"

Did this logline really had the ERROR severity? IIRC it has INFO.

> this gave following output/error
**Error: postgres2 is not a member of cluster**

Because the cluster name is **postgres-cluster**, not _postgres_

> WAL segment 0000000A000000000000001E has already been removed

Yes, the node was down for too long and the primary has already recycled this WAL segment.
You have to configure continuous archiving and PITR in order to avoid it in the future.
The second option would be increasing `wal_keep_segments`.
The third option - wait for the new release which includes https://github.com/zalando/patroni/pull/1392 and configure permanent sots.

**Comment by sysadmin-james** [comment]
yes, i made a mistake, thought postgres was the user required to reinitialise the cluster instead of clusterName "postgres-cluster". 
Use the following  command to fix this issue 
**_"2020-04-14 05:43:51,739 INFO: master: history=1	0/678BA08	no recovery target specified"_**

docker exec -it UAT_postgres2.1.z7i0uuwifizmap7rhgq900fvx   patronictl -c patroni.yml reinit postgres-cluster postgres2

+---------------------+--------------+--------------+--------+---------+----+-----------+-----------------+
|       Cluster       |    Member    |     Host     |  Role  |  State  | TL | Lag in MB | Pending restart |
+---------------------+--------------+--------------+--------+---------+----+-----------+-----------------+
| postgres-cluster    | postgres1    | postgres1 |           | running | 10 |         0 |        *        |
| postgres-cluster    | postgres2    | postgres2 |           | running | 10 |       659 |        *        |
| postgres-cluster    | postgres3    | postgres3 |   Leader  | running | 10 |         0 |        *        |
+---------------------+--------------+--------------+--------+---------+----+-----------+-----------------+
Are you sure you want to reinitialize members postgres2? [y/N]: y
Success: reinitialize for member postgres2

current Status

Tue Apr 14 07:21:01 UTC 2020
root@UAT:~# docker exec -it UAT_postgres2.1.z7i0uuwifizmap7rhgq900fvx  patronictl -c patroni.yml list
+---------------------+--------------+--------------+--------+---------+----+-----------+-----------------+
|       Cluster       |    Member    |     Host     |  Role  |  State  | TL | Lag in MB | Pending restart |
+---------------------+--------------+--------------+--------+---------+----+-----------+-----------------+
|     postgres-cluster |     postgres1 |     postgres1 |        | running | 10 |         0 |        *        |
|     postgres-cluster |     postgres2 |     postgres2 |        | running | 10 |         0 |        *        |
|     postgres-cluster |     postgres3 |     postgres3 | Leader | running | 10 |         0 |        *        |
+---------------------+--------------+--------------+--------+---------+----+-----------+-----------------+
  
I will work on your suggestions for second query.

One last thing, i didn't understand in this table why for all postgres services there is Pending restart marked,  even if i recreated all services (containers) of this cluster(I am running this cluster in Docker-swarm mode), all of them are still marked  as pending restart.

Is this something to worry about. It wasn't coming when i was testing on  brand new setup.

As far as i remember this happened/is coming only once i join  my new postgres-cluster to an existing postgres database. 

Thanks CyberDem0n
